File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
from functools import lru_cache
from math import inf

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)

        @lru_cache(None)
        def dp(i: int, holding: bool) -> int:
            if i == n:
                return 0 if not holding else -inf

            price = prices[i]

            if not holding:
                skip = dp(i + 1, False)
                buy  = -price + dp(i + 1, True)
                return max(skip, buy)
            else:
                skip = dp(i + 1, True)
                sell = price        # once sold, transaction is complete
                return max(skip, sell)

        return dp(0, False)

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
  public int maxProfit(int[] prices) {
    int l = 0;
    int r = 1;
    int profit = 0;
    while (r < prices.length) {
      if (prices[r] > prices[l]) {
        int cur = prices[r] - prices[l];
        profit = Math.max(cur, profit);
      } else {
        l = r;
      }
      r += 1;
    }
    return profit;
  }
}